### ISSUE TITLE
validation error stack unhelpful

### ISSUE THREAD

**Issue by m3l7** [balderdashy/waterline#1194]
hello,
I'm trying to get the stack trace of a validation error, after doing a model action.
However, the best I can get is error.rawStack, which gives:

      at WLValidationError.WLError (/home/dati/Sorgenti/Node/d/node_modules/sails/node_modules/waterline/lib/waterline/error/WLError.js:33:18)
      at new WLValidationError (/home/dati/Sorgenti/Node/d/node_modules/sails/node_modules/waterline/lib/waterline/error/WLValidationError.js:20:28)
      at module.exports.validate.async.series.runner (/home/dati/Sorgenti/Node/d/node_modules/sails/node_modules/waterline/lib/waterline/query/validate.js:45:43)
      at allValidationsChecked (/home/dati/Sorgenti/Node/d/node_modules/sails/node_modules/waterline/lib/waterline/core/validations.js:203:5)
      at done (/home/dati/Sorgenti/Node/d/node_modules/sails/node_modules/async/lib/async.js:132:19)
      at _toString (/home/dati/Sorgenti/Node/d/node_modules/sails/node_modules/async/lib/async.js:32:16)
      at Validator.validate (/home/dati/Sorgenti/Node/d/node_modules/sails/node_modules/waterline/lib/waterline/core/validations.js:194:14)
      at done (/home/dati/Sorgenti/Node/d/node_modules/sails/node_modules/async/lib/async.js:132:19)
      at _toString (/home/dati/Sorgenti/Node/d/node_modules/sails/node_modules/async/lib/async.js:32:16)
      at /home/dati/Sorgenti/Node/d/node_modules/sails/node_modules/waterline/lib/waterline/core/validations.js:157:64
      at async.each.callback (/home/dati/Sorgenti/Node/d/node_modules/sails/node_modules/async/lib/async.js:122:13)
      at _each (/home/dati/Sorgenti/Node/d/node_modules/sails/node_modules/async/lib/async.js:46:13)
      at Object.async.each (/home/dati/Sorgenti/Node/d/node_modules/sails/node_modules/async/lib/async.js:121:9)
      at validate (/home/dati/Sorgenti/Node/d/node_modules/sails/node_modules/waterline/lib/waterline/core/validations.js:156:11)
      at async.each.callback (/home/dati/Sorgenti/Node/d/node_modules/sails/node_modules/async/lib/async.js:122:13)
      at _each (/home/dati/Sorgenti/Node/d/node_modules/sails/node_modules/async/lib/async.js:46:13)
      at Object.async.each (/home/dati/Sorgenti/Node/d/node_modules/sails/node_modules/async/lib/async.js:121:9)
      at Validator.validate (/home/dati/Sorgenti/Node/d/node_modules/sails/node_modules/waterline/lib/waterline/core/validations.js:200:9)
      at module.exports.validate.async.series.runner (/home/dati/Sorgenti/Node/d/node_modules/sails/node_modules/waterline/lib/waterline/query/validate.js:41:25)
      at async.series.results (/home/dati/Sorgenti/Node/d/node_modules/sails/node_modules/async/lib/async.js:607:21)
      at _asyncMap (/home/dati/Sorgenti/Node/d/node_modules/sails/node_modules/async/lib/async.js:246:17)
      at async.eachSeries.iterate (/home/dati/Sorgenti/Node/d/node_modules/sails/node_modules/async/lib/async.js:146:13)

which is not very helpful. I would like to have at least the line where the model action is called.

is there a way to get a better trace?
thanks

**Comment by m3l7** [comment]
can this be related to the particular waterline adapter? (mysql in my case)

stack trace is not propagated correctly even for simple queries:

      code.user.save()
      .catch(function(err){
      	console.log(err)
      })

      > console.log(err);
      < Error (E_VALIDATION) :: 1 attribute is invalid
      <     at WLValidationError.WLError (/home/dati/Sorgenti/Node/d/node_modules/sails/node_modules/waterline/lib/waterline/error/WLError.js:33:18)
      <     at new WLValidationError (/home/dati/Sorgenti/Node/d/node_modules/sails/node_modules/waterline/lib/waterline/error/WLValidationError.js:20:28)
      <     at module.exports.validate.async.series.runner (/home/dati/Sorgenti/Node/d/node_modules/sails/node_modules/waterline/lib/waterline/query/validate.js:45:43)
      <     at allValidationsChecked (/home/dati/Sorgenti/Node/d/node_modules/sails/node_modules/waterline/lib/waterline/core/validations.js:203:5)
      <     at done (/home/dati/Sorgenti/Node/d/node_modules/sails/node_modules/async/lib/async.js:132:19)
      <     at _toString (/home/dati/Sorgenti/Node/d/node_modules/sails/node_modules/async/lib/async.js:32:16)
      <     at Validator.validate (/home/dati/Sorgenti/Node/d/node_modules/sails/node_modules/waterline/lib/waterline/core/validations.js:194:14)
      <     at done (/home/dati/Sorgenti/Node/d/node_modules/sails/node_modules/async/lib/async.js:132:19)
      <     at _toString (/home/dati/Sorgenti/Node/d/node_modules/sails/node_modules/async/lib/async.js:32:16)
      <     at /home/dati/Sorgenti/Node/d/node_modules/sails/node_modules/waterline/lib/waterline/core/validations.js:157:64
      <     at async.each.callback (/home/dati/Sorgenti/Node/d/node_modules/sails/node_modules/async/lib/async.js:122:13)
      <     at _each (/home/dati/Sorgenti/Node/d/node_modules/sails/node_modules/async/lib/async.js:46:13)
      <     at Object.async.each (/home/dati/Sorgenti/Node/d/node_modules/sails/node_modules/async/lib/async.js:121:9)
      <     at validate (/home/dati/Sorgenti/Node/d/node_modules/sails/node_modules/waterline/lib/waterline/core/validations.js:156:11)
      <     at async.each.callback (/home/dati/Sorgenti/Node/d/node_modules/sails/node_modules/async/lib/async.js:122:13)
      <     at _each (/home/dati/Sorgenti/Node/d/node_modules/sails/node_modules/async/lib/async.js:46:13)
      <     at Object.async.each (/home/dati/Sorgenti/Node/d/node_modules/sails/node_modules/async/lib/async.js:121:9)
      <     at Validator.validate (/home/dati/Sorgenti/Node/d/node_modules/sails/node_modules/waterline/lib/waterline/core/validations.js:200:9)
      <     at module.exports.validate.async.series.runner (/home/dati/Sorgenti/Node/d/node_modules/sails/node_modules/waterline/lib/waterline/query/validate.js:41:25)
      <     at async.series.results (/home/dati/Sorgenti/Node/d/node_modules/sails/node_modules/async/lib/async.js:607:21)
      <     at _asyncMap (/home/dati/Sorgenti/Node/d/node_modules/sails/node_modules/async/lib/async.js:246:17)
      <     at async.eachSeries.iterate (/home/dati/Sorgenti/Node/d/node_modules/sails/node_modules/async/lib/async.js:146:13)
      < Invalid attributes sent to User:
      <  • points
      <    • `undefined` should be a integer (instead of "NaN", which is a number)


this seems to me wrong, since I need to wrap everytime a waterline error inside another error to propagate the stack.

**Comment by m3l7** [comment]
I've just tried with mongodb adapter, and it has the same behaviour.

**Comment by m3l7** [comment]
I want to point out thaat this is not related to the specific query code.user.save().
It happens to every db query

**Comment by sailsbot** [comment]
Thanks for posting, @m3l7.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by particlebanana** [comment]
I'm not sure what the question is here. The validation will be called inside of Waterline so the stack trace will always be the same. Closing this.

**Comment by m3l7** [comment]
yeah, that's correct but it's quite disappointing since I can't trace where the waterline function is being called in my code.

I'm wondering how are the best practices to track waterline validation errors then.

The only solution I'm thinking is to catch every waterline call and rethrow the error in case of validation error, but it's quite annoying
